engine/src/tilecachegl3.x.cpp

```diff
@@ -301,27 +540,80 @@ bool MCTileCacheOpenGLCompositor_AllocateTile(void *p_context, int32_t p_size, c
 	MCTileCacheOpenGLCompositorContext *self;
 	self = (MCTileCacheOpenGLCompositorContext *)p_context;
 
-	// If the stride is exactly one tile wide, we don't need a copy.
-	void *t_data;
-	t_data = nil;
-	if (p_stride == p_size * sizeof(uint32_t))
-		t_data = (void *)p_bits;
-	else if (MCMemoryAllocate(p_size * p_size * sizeof(uint32_t), t_data))
-	{
-		// Copy across each scanline of the tile into the buffer.
-		for(int32_t y = 0; y < p_size; y++)
-			memcpy((uint8_t *)t_data + y * p_size * sizeof(uint32_t), (uint8_t *)p_bits + p_stride * y, p_size * sizeof(uint32_t));
-	}
-	
 	void *t_tile;
-	uint32_t t_tile_id;
 	t_tile = nil;
-	if (t_data != nil && MCTileCacheOpenGLCompositorCreateTile(self, t_tile_id))
+	uint32_t t_tile_id;
+	if (!MCTileCacheOpenGLCompositorCreateTile(self, t_tile_id))
+		return false;
+
+	// Fetch the super/sub indices of the tile.
+	uint32_t t_super_tile_index, t_sub_tile_index;
+	MCTileCacheOpenGLCompositorDecodeTile(self, t_tile_id, t_super_tile_index, t_sub_tile_index);
+	
+	uint32_t t_tile_count = kSuperTileSize / self->tile_size;
+
+	// Calculate its location.
+	GLint t_x, t_y;
+	t_x = t_sub_tile_index % t_tile_count;
+	t_y = t_sub_tile_index / t_tile_count;
+	
+	super_tile *t_super_tile;
+	t_super_tile = self->super_tiles[t_super_tile_index];
+
+	bool t_have_hw_buffer = false;
+
+#if defined(MCANDROIDUSEHARDWAREBUFFER)
+	t_have_hw_buffer = t_super_tile->hw_buffer != nil;
```

Comment:
Similar to the above rather than use #ifdef here - instead make a higher-level method which fails if no hardware buffer. i.e.
if (!MCTileCacheHardwareBufferUpdate(super_tile, p_size, p_bits, p_stride, p_x, p_y))
{
   ... slow non-hardware buffer path ...
}
```

Reply:
Fixed in updated commits